### ISSUE TITLE
Queue stop not working as expected.

### ISSUE THREAD

**Issue by odarriba** [sorentwo/oban#318]
### Environment

- Oban Version: Oban v2.0.0 - Oban.Web v2.0.0 - Oban.Pro v0.3.0
- PostgreSQL Version 12 (Heroku)
- Elixir & Erlang/OTP Versions (`elixir --version`)

```
Erlang/OTP 22 [erts-10.7.1] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [hipe]

Elixir 1.10.4 (compiled with Erlang/OTP 21)
```

### Current Behavior

We have a setup of two nodes on Heroku (non clustered, as Heroku doesn´t support that), both running Oban with a `default` queue.

From the Oban Web, we pause the queue. it looks like paused on the UI, but the jobs continue beign executed, and if we reload the Oban Web dashboard, it appears as started again.

### Expected Behavior

We expect that the queue stop is executed and maintained until the queue is started again.

**Comment by sorentwo** [comment]
In my tests locally, on the demo site, and in a production app the pause functionality works and the pause persists between reloads. A few questions to try and narrow things down:

* Is this reproducible and do you have any errors logged from around the same time?
* How long was it between when you paused and then reloaded to look again?
* Was there a deploy or an application restart in-between? The pause only persists until the node is restarted. If you paused and then reloaded after a new deployment then the queues would have started back up.

Regarding the jobs continuing to execute, when you pause it stops fetching new jobs. Any currently executing jobs will continue unless you cancel them.

**Comment by odarriba** [comment]
Hi,

Thanks for your response.

### Is this reproducible and do you have any errors logged from around the same time?
Not that we could see on logs or AppSignal. I have just tried again right now and the behaviour persists - no error or message on the log.

However, we are only seeing this behaviour on production - staging is working fine with the same environment (two Heroku nodes + Heroku PostgreSQL DB). Could it be something broken on the DB?

### How long was it between when you paused and then reloaded to look again?
It doesn´t make a difference, I think. I have tried reloading just after clicked the stop action, or waiting like a minute or so - no changes.

### Was there a deploy or an application restart in-between?
No deploy or daily restart of the app.

We expect the tasks that are already running to continue, but once we click on stop we can enqueue new tasks and they are processed at the moment - so the queue is active.

I will try to set oban on debug mode to see if there is something related to the DB on that environment - however, at the moment no error is logged.

**Comment by odarriba** [comment]
I think we have detected the issue.

We are using `pg_bouncer` on production to control memory needs (we currently use schema-based multi tenancy, causing connections on server side to use LOTS of memory).

It looks like transaction pooling on `pg_bouncer` doesn't support `listen` or `notify` functionality of PostgreSQL, and that's what Oban is using to notify workers to stop certain queue.

Is there any way to force Oban to use another Repo just for the control and internal connections? (not the ones used by the workers to execute code).

Is there any other side-effect we could be suffering but not detecting?

**Comment by sorentwo** [comment]
> It looks like transaction pooling on `pg_bouncer` doesn't support `listen` or `notify` functionality of PostgreSQL, and that's what Oban is using to notify workers to stop certain queue.

That totally makes sense. I haven't encountered `pg_bouncer` in transaction pooling mode before and had no clue.

> Is there any way to force Oban to use another Repo just for the control and internal connections? (not the ones used by the workers to execute code).

Notifications are used for all of the multi-node functionality (start, stop, pause, scale, kill job, etc). Notify events also used to trigger job execution outside of the polling schedule, so the notification connection must be to the same database. There isn't any way to work around that currently.

To do that we'd need to add another option to the `Oban.Config` struct to support an alternative Repo config.

> Is there any other side-effect we could be suffering but not detecting?

As I noted above, all multi-node commands won't work without pubsub functionality. If you're locked in to using transaction pooling we'll need to work out some alternative configuration.

**Comment by odarriba** [comment]
We have made a small test using pg_bouncer on one of our pre-prod environments, and that was definitively the root of the issue.

We have also created another `Repo` which connects directly to the database, and used it on Oban setting it in the `repo` configuration value, so Oban and Oban.Web will use that repo for notifications, listen for events, etc.

That is a solution that can work for us, as the memory issues in PostgreSQL we are facing are related to multiple (100+) schemas, which are not used by Oban.

Also, the jobs insertion is made using the regular repo on the application, so it goes through `pg_bouncer`.

That configuration seems to work fine, but we will do further testing during next couple of days.
Maybe this is a possible solution for other users with the same issue 😅
